server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4181,10 +4183,25 @@ public void waitForSegmentGeneration(long segmentGeneration, ActionListener<Long
         waitForPrimaryTermAndGeneration(getOperationPrimaryTerm(), segmentGeneration, listener);
     }
 
+    private final SubscribableListener<Void> waitForPrimaryTermAndGenerationListeners = new SubscribableListener<Void>();
```

Comment:
nit: something like `waitForStartedOrClosedShard` seems more of a fitting name. I think the second `Void` could be omitted. Also why not move this up to where the other fields are?